lib/git_worktree.rb

```diff
@@ -0,0 +1,336 @@
+    entry = find_entry(path, commit_sha)
+    entry && entry[:type] == :tree
+  end
+
+  def read_file(path, commit_sha = nil)
+    read_entry(fetch_entry(path, commit_sha))
+  end
+
+  def read_entry(entry)
+    @repo.lookup(entry[:oid]).content
+  end
+
+  def directory?(path, commit_sha = nil)
+    entry = fetch_entry(path, commit_sha)
+    entry[:type] == :tree
+  end
+
+  def entries(path, commit_sha = nil)
+    tree = get_tree(path, commit_sha)
+    tree.find_all.collect { |e| e[:name] }
+  end
+
+  def nodes(path, commit_sha = nil)
+    tree = path.empty? ? lookup_commit_tree(commit_sha || @commit_sha) : get_tree(path, commit_sha)
+    entries = tree.find_all
+    entries.each do |entry|
+      entry[:full_name] = File.join(@base_name, path, entry[:name])
+      entry[:rel_path] = File.join(path, entry[:name])
+    end
+  end
+
+  def save_changes(message, owner = :local)
+    cid = commit(message)
+    if owner == :local
+      lock { merge(cid) }
+    else
+      merge_and_push(cid)
+    end
+    true
+  end
+
+  def commit(message)
+    tree = @current_index.write_tree(@repo)
+    parents = @repo.empty? ? [] : [@repo.ref(HEAD_REF).target].compact
+    create_commit(message, tree, parents)
+  end
+
+  def file_attributes(fname)
+    walker = Rugged::Walker.new(@repo)
+    walker.sorting(Rugged::SORT_DATE)
+    walker.push(@repo.ref(HEAD_REF).target)
+    commit = walker.find { |c| c.diff(:paths => [fname]).size > 0 }
+    return {} unless commit
+    {:updated_on => commit.time.gmtime, :updated_by => commit.author[:name]}
+  end
+
+  def merge(commit, rebase = false)
+    master_branch = @repo.ref(HEAD_REF)
+    merge_index = master_branch ? @repo.merge_commits(master_branch.target, commit) : nil
+    if merge_index && merge_index.conflicts?
+      result = differences_with_master(commit)
+      raise MiqException::MiqGitConflicts, result
+    end
+    commit = rebase(commit, merge_index, master_branch ? master_branch.target : nil) if rebase
+    @repo.reset(commit, :soft)
+  end
+
+  def rebase(commit, merge_index, parent)
+    commit_obj = commit if commit.class == Rugged::Commit
+    commit_obj ||= @repo.lookup(commit)
+    Rugged::Commit.create(@repo, :author    => commit_obj.author,
+                                 :committer => commit_obj.author,
+                                 :message   => commit_obj.message,
+                                 :parents   => parent ? [parent] : [],
+                                 :tree      => merge_index.write_tree(@repo))
+  end
+
+  def merge_and_push(commit)
+    rebase = false
+    push_lock do
+      @saved_cid = @repo.ref(HEAD_REF).target.oid
+      merge(commit, rebase)
+      rebase = true
+      @repo.push(@remote_name, [HEAD_REF], :credentials => @cred)
+    end
+  end
+
+  def fetch
+    options = {:credentials => @cred}
+    ssl_no_verify_options(options) do
+      @repo.fetch(@remote_name, options)
+    end
+  end
+
+  def pull
+    lock { fetch_and_merge }
+  end
```

Comment:
@Fryguy 
Fetch_merge is not part of rugged, instead they provide a fetch function and  commit and merge_commits. In our case we have to do the rebasing and reapplying the commits if there is a conflict.